libsplinter/src/rest_api/auth/roles/store/mod.rs

```diff
@@ -0,0 +1,379 @@
+impl Role {
+    /// Returns the role's id.
+    pub fn id(&self) -> &str {
+        &self.id
+    }
+
+    /// Returns the role's display name.
+    pub fn display_name(&self) -> &str {
+        &self.display_name
+    }
+
+    /// Returns the role's permissions.
+    pub fn permissions(&self) -> &[String] {
+        &self.permissions
+    }
+
+    /// Convert this role back into a builder, in order to update its values.
+    pub fn into_update_builder(self) -> RoleUpdateBuilder {
+        RoleUpdateBuilder {
+            id: self.id,
+            display_name: Some(self.display_name),
+            permissions: self.permissions,
+        }
+    }
+}
+
+/// A builder to create new roles.
+#[derive(Default)]
+pub struct RoleBuilder {
+    id: Option<String>,
+    display_name: Option<String>,
+    permissions: Vec<String>,
+}
+
+impl RoleBuilder {
+    /// Constructs a new builder.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Sets the ID for the new role.
+    pub fn with_id(mut self, id: String) -> Self {
+        self.id = Some(id);
+        self
+    }
+
+    /// Sets the display name for the new role.
+    pub fn with_display_name(mut self, display_name: String) -> Self {
+        self.display_name = Some(display_name);
+        self
+    }
+
+    /// Sets the permissions for the new role.
+    pub fn with_permissionss(mut self, permissions: Vec<String>) -> Self {
+        self.permissions = permissions;
+        self
+    }
+
+    /// Builds the new Role.
+    ///
+    /// # Errors
+    ///
+    /// Returns an [`InvalidStateError`] under the following conditions:
+    /// * no ID or an empty ID was provided
+    /// * no display name or an empty display name was provided
+    /// * empty permissions were provided
+    pub fn build(self) -> Result<Role, InvalidStateError> {
+        if self.permissions.is_empty() {
+            return Err(InvalidStateError::with_message(
+                "A role requires at least one permission".into(),
+            ));
+        }
+
+        let id = self
+            .id
+            .ok_or_else(|| InvalidStateError::with_message("A role requires an id field".into()))?;
+        if id.is_empty() {
+            return Err(InvalidStateError::with_message(
+                "A role requires a non-empty id field".into(),
+            ));
+        }
+        let display_name = self.display_name.ok_or_else(|| {
+            InvalidStateError::with_message("A role requires a display_name field".into())
+        })?;
+
+        if display_name.is_empty() {
+            return Err(InvalidStateError::with_message(
+                "A role requires a non-empty display_name field".into(),
+            ));
+        }
+
+        Ok(Role {
+            id,
+            display_name,
+            permissions: self.permissions,
+        })
+    }
+}
+
+/// Updates an existing role.
+///
+/// This builder only allows the updatable fields to be modified.
+pub struct RoleUpdateBuilder {
+    id: String,
```

Comment:
Should these all be Option<> (its a builder).

Reply:
In the update builder, only certain fields can be modified. Id is not one of them.